### PR TITLE
[microTVM][ARM] Keep microtvm testing only in QEMU Image

### PR DESCRIPTION
**Pull request by mehrdadh**
This would prevent rebuilding multiple image on each update on CMSIS, ethosu, etc.
cc @Mousius @alanmacd @areusch @gromero @leandron

### CODE REVIEW

**Review by areusch — COMMENTED** [review]
let's see what @u99127 @Mousius @manupa-arm have to say here first. i think this is mainly to make it operationally easier for us to ensure the microTVM upstream tests use the same deps as we install in Reference VM. i think that as we get towards supporting `tvmc` in a release and doing further testing with microTVM against `tvmc` in its own venv with e.g. Zephyr installed separately, we could probs focus on that in place of Reference VM. anyway, the ci_ images remain the test authority, so this is again mainly an operational convenience for folks using Reference VM as we do to test on-device at OctoML.

File: docker/Dockerfile.ci_cpu [hunk 1]
```diff
@@ -110,10 +110,6 @@ RUN bash /install/ubuntu_install_boost.sh
 COPY install/ubuntu_install_caffe.sh /install/ubuntu_install_caffe.sh
 RUN bash /install/ubuntu_install_caffe.sh
 
-# Github Arm(R) Ethos(TM)-N NPU driver
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply docker/Dockerfile.ci_cpu]: If my understanding is correct and this is used for running this command:
```
run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-contrib-test_ethosu tests/python/contrib/test_ethosu -n auto
```
I suggest we also move this to ci_qemu to have all microtvm tests in one place.

**Comment by manupa-arm** [comment]
In order to reason about this, it would be great to understand why do we maintain two seperate containers for microTVM and TVM, especially when QEMU is a software that runs on the cpu. (Sorry I might've missed the discussion why went on to create to two seperate images)

In an alternate world, we could consolidate all testing dependencies to ci_cpu and keep TVM and microTVM aligned in-terms of testing environment, unless we need conflicting dependencies in the testing for TVM vs microTVM *not just additive ones*.

**Comment by mehrdadh** [comment]
Adding to Andrew's point, I think we can remove `USE_MICRO` since it's almost always ON. Also, I prefer to keep ci_qemu separate and also call it ci_microtvm. I believe this image could grow larger in future and if we combine it with ci_cpu it becomes similar to ci_cpu.

**Comment by Mousius** [comment]
> Adding to Andrew's point, I think we can remove `USE_MICRO` since it's almost always ON. Also, I prefer to keep ci_qemu separate and also call it ci_microtvm. I believe this image could grow larger in future and if we combine it with ci_cpu it becomes similar to ci_cpu.

I support this approach as well, having two smaller containers with a specific responsibility rather than one monolithic container should make it easier to manage in the long term as the needs evolve. The only downside is the typical container hopping for development, though I think we can figure out how to get that to work also.

Minor note, I think `ci_micro` rather than `ci_microtvm` as it's already identified as a `tvm` image?

**Comment by manupa-arm** [comment]
I agree that having modular (smaller) container aids the ease of reproducability from a developer PoV.

I'm worried with this approach is that this creates a room for set of diveregent dependencies to be introduced between microTVM vs TVM (the deps that are not specific to micro), especially when they are not seperate projects. Right now, only thing that keeps the common deps align is on the basis that we use the same helper scripts (ubuntu_install*.sh) to construct the image.

E.g. : This could open up the possibility to advance the version of tensorflow supported by TVM project while microTVM lacking behind because there will now be tested in two environments.

Maybe we could create a base container with such common deps to ensure such divergence are harder to do ? (and extend that image into ci_cpu and ci_qemu)

**Comment by manupa-arm** [comment]
cc : @leandron

**Comment by leandron** [comment]
Looking at the alternatives being proposed until now, I feel that for *our current type of usage* is makes more sense to unify `ci_qemu` into `ci_cpu`, for two reasons:

1. [According to our Jenkinsfile](https://github.com/apache/tvm/blob/ce77d20c4615c655214d0bd38307813e539ba9a2/Jenkinsfile#L503-L511), both jobs that require `ci_cpu` and `ci_qemu` will run on the same target machines, so the expectation is that our Jenkins nodes will more commonly pull `ci_cpu + ci_qemu` in terms of storage
2. Creating a base image with common dependencies, would also generate a side problem, which is a _time dependency_ to rebuild our images (we can't build all of them at the same time), and the need to create a generic label e.g. `latest` or `last-successful` for this base image to be picked up in the dependant images. I'd like to avoid this if possible.

Given our current Docker update practices that don't require updating all images at once, I feel we better to consolidate the dependencies in one single image until we can devise better rules for Docker updating.

**Comment by areusch** [comment]
i think while i'm sympathetic to reducing the overall image disk size, in practice:
- DockerHub stores these for us and we don't have a concern there
- we have the problem of docker image bloat hogging disk on our CI nodes anyway due to image revisions, and have already solved that.

so i don't think we need to make that a priority in this decision. i think then that it comes down to what's operationally easier, and i think that's the POV @Mousius and @manupa-arm were advocating. i tend to agree here--another way that we could get into trouble with consolidating `ci_cpu` and `ci_qemu` is two different `docker/install` scripts could add conflicting packages. we can mitigate that by keeping microTVM separated, and I'd vote for continuing that with renaming `ci_qemu` to `ci_micro`.

cc @driazati @konturn in case they have other thoughts from a CI perspective.

**Comment by mehrdadh** [comment]
I'm in favor of having a base image with common dependencies to eliminate human errors in build process for each docker image, but even with that people might add any script to any of the docker files. So as a reviewer we need to keep an eye on changes on those files. Also I think there shouldn't be any divergence between TVM and microTVM dependencies and I would like to second @manupa-arm's point on this.
For now, I think we can try to keep all microtvm testing in a single docker image and rename it to ci_micro to it make more readable. We can think about other proposals and maybe discuss it in one of the TVM community meetings.

**Comment by manupa-arm** [comment]
Thanks all for the discussion!.

> we can avoid dependency drift by being careful in reviewing Docker image changes.

If we are solely relying on the reviewers, we need would need better written guidance in an OSS project to aid both reviewers (to cite) and authors (to follow), on what could be added to : ci_base (if we are taking this approach), ci_micro and ci_cpu. 
NOTE : Im scoping this discussion to the specific issue being discussed here.

> but even with that people might add any script to any of the docker files. So as a reviewer we need to keep an eye on changes on those files.

I suppose how ci_base would help here is that we can make the above guidance to be simplified to say all the layers in the image added by ci_micro Dockerfile has to be micro specific.

**Comment by mehrdadh** [comment]
@manupa-arm I agree with you. Both better documentation and having a ci_base image which has common dependencies are great ideas.
It would be great if we can make an action items list based on the discussion here.

**Comment by areusch** [comment]
i think most of the discussion here has aligned around purpose-built CI images. @leandron are your concerns assuaged? 

regarding `ci_base`: I'm somewhat supportive of this, but currently unsure whether this would move the needle very much in practice and there are some operational concerns. we currently build images mostly automatically around the same time via the nightly rebuild and also in CI. the common deps we're talking about here are mostly a mix of `apt` packages plus some things built from sources downloaded from (i think) version-pinned URLs. there are some other things in here, to be sure--rust packages, for instance. i do agree that `ci_base` would enforce that we don't diverge, but just stating that i don't know, given the above, that i expect to see a huge effect in creating that. i agree with @Mousius that the layers would ideally be reused; i wouldn't underestimate the network transfer time, however.

I do want to note that we also mitigate drift in other was e.g. via common install scripts and (soon) via locked python dependencies (this more addresses the example of tensorflow divergence). however, i acknowledge the drawback that this is merely convention and dependent on reviewers.

for verticals like microTVM, i think a separate image makes sense. for e.g. backends, it's less clear--what if you want to test cuda with paddlepaddle? there is no such paddlepaddle dep for arm64, so it can't go in the base image. does this mean base images need to be arch-specific--if so, where's the line (e.g. what if it's available for 20.04 but not 18.04)?

so far as action items, i think we should capture the need of either a common base image or documentation into a GH issue. we should also update our docs accordingly. would love to hear more thoughts @Mousius @manupa-arm @leandron @driazati

File: docker/Dockerfile.ci_cpu [hunk 2]
```diff
@@ -124,15 +124,6 @@ RUN bash /install/ubuntu_install_androidsdk.sh
 ENV ANDROID_HOME=/opt/android-sdk-linux/
 ENV ANDROID_NDK_HOME=/opt/android-sdk-linux/ndk/21.3.6528147/
 
-# Install CMSIS_NN
-COPY install/ubuntu_install_cmsis.sh /install/ubuntu_install_cmsis.sh
-RUN bash /install/ubuntu_install_cmsis.sh /opt/arm/ethosu/cmsis
-ENV CMSIS_PATH=/opt/arm/ethosu/cmsis/
-
-# Arm(R) Ethos(TM)-U NPU driver
-COPY install/ubuntu_install_ethosu_driver_stack.sh /install/ubuntu_install_ethosu_driver_stack.sh
```
Comment (manupa-arm) [inline docker/Dockerfile.ci_cpu:133]: run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-contrib-test_ethosu tests/python/contrib/test_ethosu -n auto

we need this line moved alongside as well.

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Im just suggesting a mandatory change orthogonal to the conversation.

File: docker/Dockerfile.ci_cpu (hunk 2 above)
Comment (mehrdadh) [reply docker/Dockerfile.ci_cpu:133]: we should move that after we updated the qemu image with the dependencies for this test

Comment (mehrdadh) [reply docker/Dockerfile.ci_cpu:133]: I can update this PR by adding `Arm(R) Ethos(TM)-U NPU driver` to the qemu, we merge this. Then we will update the qemu image and then send another PR to remove `test_ethosu` and move it to qemu.

Comment (mehrdadh) [reply docker/Dockerfile.ci_cpu:133]: sorry we already have that. I meant `Github Arm(R) Ethos(TM)-N NPU driver`, I believe that is required, correct me if I'm wrong.

Comment (mehrdadh) [reply docker/Dockerfile.ci_cpu:133]: I tested it in qemu, looks fine. So I moved it

**Comment by areusch** [comment]
@manupa-arm can you take a look again?

**Review by manupa-arm — COMMENTED** [review]
Thanks! looking good.

It would be great to capture the steps discussed and agreed in the issue (as per @areusch) and link to the PR.

@ashutosh-arm could you take a look as well from cmsis tests, whether they would be all run in ci_qemu now ?

**Comment by ashutosh-arm** [comment]
> Thanks! looking good.
> 
> It would be great to capture the steps discussed and agreed in the issue (as per @areusch) and link to the PR.
> 
> @ashutosh-arm could you take a look as well from cmsis tests, whether they would be all run in ci_qemu now ?

afaik CMSIS-NN tests do not run on ci_qemu at present. In this code change, I don't see CMSIS-NN tests moved to tests/scripts/task_python_microtvm.sh. I am referring to https://github.com/apache/tvm/blob/6eb3a1fc3688bede69efd7f14b313b35497ccf02/tests/scripts/task_python_integration.sh#L64. Am I missing something @mehrdadh / @manupa-arm ?

**Comment by manupa-arm** [comment]
I see. 

Yes none of the contrib tests are run (that is how cmsis tests are run today in ci_cpu).
We need to add that as well then.

**Comment by mehrdadh** [comment]
@ashutosh-arm thanks for catching this. I moved test_cmsisnn tests to qemu, there's also test_ethosn which we should move, but we need to update qemu image for that.
I think I need to send another PR for that along with sharding the qemu tests since it has timeout issue right now

**Comment by manupa-arm** [comment]
@mehrdadh test_ethosn (unlike test_ethosu) is not related to microTVM.

**Comment by mehrdadh** [comment]
@manupa-arm thanks for clarification! Then I think this PR is ready, I just need to figure out the timeout issue.
cc @areusch @driazati

**Comment by areusch** [comment]
just musing about naming--previously we called this image `ci_qemu` because it contained qemu. it seems like with this effort + riscv32 + the hexagon effort, we are gravitating more towards purpose-driven CI images (we discussed this before). there is already a ci_arm--shall we consider a container rename such as:
- `ci_arm` -> `ci_arm_aarch64`
- `ci_qemu` -> `ci_arm_cortexm`

**Comment by mehrdadh** [comment]
@areusch I like to change the name, but I think ci_arm_cortexm is too specific. We could have riscv support with Zephyr in future which I assume the testing is gonna be part of this image. Then it would be confusing. I think ci_micro works better here. wdyt?

**Comment by u99127** [comment]
> just musing about naming--previously we called this image `ci_qemu` because it contained qemu. it seems like with this effort + riscv32 + the hexagon effort, we are gravitating more towards purpose-driven CI images (we discussed this before). there is already a ci_arm--shall we consider a container rename such as:
> 
> * `ci_arm` -> `ci_arm_aarch64`
> * `ci_qemu` -> `ci_arm_cortexm`

I'd suggest that ci_arm is renamed to ci_aarch64  . 

Ramana

**Comment by areusch** [comment]
@u99127 i'm fine with that, should we then `ci_cortexm` instead?

**Review by areusch — APPROVED** [review]

**Comment by areusch** [comment]
@manupa-arm please take another look when you have a minute

**Comment by mehrdadh** [comment]
just rebased with main, timeout issue should be resolved.

**Comment by mehrdadh** [comment]
@manupa-arm friendly reminder about this PR. thanks!

**Review by manupa-arm — APPROVED** [review]
Thanks! LGTM!

**Pull request merged**